logstash-core/lib/logstash/pipeline.rb

```diff
@@ -94,60 +112,164 @@ def filters?
-      flush_filters_to_output!(:final => true)
+      inflight_batches_synchronize { set_current_thread_inflight_batch(nil) }
     end
+  end
 
-    shutdown_outputs
-    wait_outputs
+  def take_batch(batch_size, batch_delay)
+    batch = []
+    # Since this is externally synchronized in `worker_look` wec can guarantee that the visibility of an insight batch
+    # guaranteed to be a full batch not a partial batch
+    set_current_thread_inflight_batch(batch)
+
+    shutdown_received = false
+    batch_size.times do |t|
+      event = t==0 ? @input_queue.take : @input_queue.poll(batch_delay)
+      # Exit early so each thread only gets one copy of this
+      # This is necessary to ensure proper shutdown!
+      next if event.nil?
+
+      if event == LogStash::SHUTDOWN
+        shutdown_received = true
+        break
+      end
 
-    @logger.info("Pipeline shutdown complete.")
-    @logger.terminal("Logstash shutdown completed")
+      batch << event
+    end
 
-    # exit code
-    return 0
-  end # def run
+    [batch, shutdown_received]
+  end
 
-  def wait_inputs
-    @input_threads.each(&:join)
+  def filter_batch(batch)
+    batch.reduce([]) do |acc,e|
+      if e.is_a?(LogStash::Event)
+        filtered = filter_func(e)
+        filtered.each {|fe| acc << fe unless fe.cancelled?}
+      end
+      acc
+    end
+  rescue Exception => e
+    # Plugins authors should manage their own exceptions in the plugin code
+    # but if an exception is raised up to the worker thread they are considered
+    # fatal and logstash will not recover from this situation.
+    #
+    # Users need to check their configuration or see if there is a bug in the
+    # plugin.
+    @logger.error("Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash.",
+                  "exception" => e, "backtrace" => e.backtrace)
+    raise
   end
 
-  def shutdown_filters
-    @flusher_thread.kill
-    @input_to_filter.push(LogStash::SHUTDOWN)
+  # Take an array of events and send them to the correct output
+  def output_batch(batch)
+    batch.reduce(Hash.new { |h, k| h[k] = [] }) do |outputs_events, event|
+      # We ask the AST to tell us which outputs to send each event to
+      output_func(event).each do |output|
+        outputs_events[output] << event
+      end
+      outputs_events
+    end.each do |output, events|
+      # Once we have a mapping of outputs => [events] we can execute them
+      output.multi_handle(events)
+    end
+  end
+
+  def set_current_thread_inflight_batch(batch)
+    @inflight_batches[Thread.current] = batch
   end
 
-  def wait_filters
-    @filter_threads.each(&:join) if @filter_threads
+  def inflight_batches_synchronize
+    @input_queue_pop_mutex.synchronize do
+      yield(@inflight_batches)
+    end
   end
 
-  def shutdown_outputs
-    # nothing, filters will do this
-    @filter_to_output.push(LogStash::SHUTDOWN)
+  def dump_inflight(file_path)
```

Comment:
I left this here because one of our goals was the ability to have a safe shutdown that could recover in flight messages. I thought it made sense to leave this in the commit to be exposed by a later patch.

Should I pull it out?